Iclass Legacy Raw Key Recovery Function

Antiklesys opened the pull request:
Based on the work described in Dismantling iClass whitepaper. 

- `hf iclass legbrute` is tested working on sample data generated accordingly to the hash0 logic.

- `hf iclass legrec ` is partially working: 
1. Logic of operations and sequence are in order 
2. Individual functionalities have been tested on sample/simulated data and was found to be effective and correct.
3. The privilege escalation operation is still NOT working during my tests (may be a silicon issue tho), but the logic in place should be correct.

Antiklesys reviewed (commented):
Forgot to add one file

iceman1001 commented on armsrc/iclass.c:
```diff
@@ -2152,3 +2152,216 @@ void iClass_Restore(iclass_restore_req_t *msg) {
     }
 }
+
+void generate_single_key_block_inverted(const uint8_t startingKey[PICOPASS_BLOCK_SIZE], uint32_t index, uint8_t keyBlock[PICOPASS_BLOCK_SIZE]) {
```
use `const uint8_t *starting_key`   instead of arrays with defined sizes.

Antiklesys replied:
Fixed

iceman1001 commented on armsrc/iclass.c:
```diff
@@ -2152,3 +2152,216 @@ void iClass_Restore(iclass_restore_req_t *msg) {
+    }
+}
+// Function to convert an unsigned int to binary string
+void intToBinary(unsigned int num, char *binaryStr, int size) {
```
we use  `uintX_t`  to denote unsigned it.  
And functions like these should be reused and located in utils.c

Antiklesys replied:
Fixed unsigned int, will check for utils.c

iceman1001 commented on armsrc/iclass.c:
```diff
@@ -2152,3 +2152,216 @@ void iClass_Restore(iclass_restore_req_t *msg) {
+}
+
+// Function to convert a binary string to hexadecimal
+uint8_t binaryToHex(char *binaryStr) {
```
see previous comment,   see if you can reuse functions from utils.c  or  commonutils..c

Antiklesys replied:
Fixed unsigned int, will check for utils.c

iceman1001 commented on armsrc/iclass.c:
```diff
@@ -2152,3 +2152,216 @@ void iClass_Restore(iclass_restore_req_t *msg) {
+}
+
+// Function to convert an unsigned int to an array of hex values
+void convertToHexArray(unsigned int num, uint8_t *partialKey) {
```
see previous comment,   see if you can reuse functions from utils.c  or  commonutils..c

iceman1001 commented on armsrc/iclass.c:
```diff
@@ -2152,3 +2152,216 @@ void iClass_Restore(iclass_restore_req_t *msg) {
+        { 0x00, 0x00, 0x00, 0x00, 0xE2, 0xD5, 0x69, 0xE9 }  //Expected mac when last 3 bits of each byte are: 111
+    };
+    //Viewing the weak macs table card 24 bits (3x8) in the form of a 24 bit decimal number
+    static uint32_t iClass_Mac_Table_Bit_Values[8] = {0, 2396745, 4793490, 7190235, 9586980, 11983725, 14380470, 16777215};
```
I am just curious how this decimal number goes together with the byte arrays above

Antiklesys replied:
The array above is a series of weak macs, those weak macs correspond to the different combinations of the last 3 bits of each key byte.
If we concatenate the last three bits of each key byte, we have a 24 bits long binary string.
If we convert that string to decimal we obtain those decimal numbers.
Xorring the index of iterations against those decimal numbers allows us to retrieve the what was the corresponding sequence of bits of the original key in decimal format.

iceman1001 replied:
Aha,   a bit complex.    If you could comment that with like you just wrote then it would make it better

iceman1001 commented on armsrc/iclass.c:
```diff
@@ -2152,3 +2152,216 @@ void iClass_Restore(iclass_restore_req_t *msg) {
+    while (bits_found == -1){
+
+    //Step3 Calculate New Key
+    uint8_t GenKeyBlock[PICOPASS_BLOCK_SIZE];
```
please use lower case when choosing variable names.   We don't use mixed case.

iceman1001 commented on client/src/cmdhficlass.c:
```diff
@@ -3842,6 +3842,304 @@ void picopass_elite_nextKey(uint8_t* key) {
+static int CmdHFiClassRecover(uint8_t key[8]) {
+
+    uint32_t payload_size = sizeof(iclass_recover_req_t);
+    uint8_t aa2_standard_key[PICOPASS_BLOCK_SIZE] = {0xFD, 0xCB, 0x5A, 0x52, 0xEA, 0x8F, 0x30, 0x90};
```
would be better to reuse the keys array in the top of this file for this purpose.

iceman1001 commented:
Nice one!  
The missing attack vector from heart of darkness paper.    I like!

iceman1001 commented on armsrc/util.c:
```diff
@@ -410,6 +410,8 @@ uint8_t binaryToHex(char *binaryStr) {
     return (uint8_t)strtoul(binaryStr, NULL, 2);
 }
 
+#define PICOPASS_BLOCK_SIZE 8
```
This define already exists in: 
`include/iclass_cmd.h`

Antiklesys replied:
I just removed the definition and replaced it with an 8 for simplicity.

iceman1001 commented:
![image](https://github.com/user-attachments/assets/74721168-1987-41f1-bccc-11eaad86acd1)

The log level is `ERR`

Antiklesys commented:
@iceman1001 whatever is in util.c for now I've just moved there for the sake of compiling.
As you mentioned there may be already some functions that could be reused.
I'll check on those and eventually that'll fix the need to redefine or not PICOPASS_BLOCK_SIZE in there.
I made all the other small fixes in the meantime, thank you for the help with the ERR/ERROR loglevel that was causing compilation outside Proxspace to fail.

iceman1001 commented:
Make sure you apply the  array parameters as pointers everywhere ,  I only pointed it out in one place.

When you feel its ready,  let me know and I will merge it.

Antiklesys commented:
@iceman1001 I cleaned up the code and removed a few trailing issues/mistakes/work in progress comments.
I reviewed also the functions to convert bin 2 hex in iclass.c but they're not doing quite the same thing as the ones I implemented are doing, as mines are also dividing a 24 bit binary string in groups of 3 bits while doing the conversion and preparation for the key.
I'm sure there may be further code changes/improvements once I get my hands on a suitable card to test this functionality inside out. I'll definitely continue working on optimizing it and improving it on my end as I come across new findings on this topic.
Also may need to remove/reduce some of the Dbprintf statements in the future depending on "how fast" this goes and to eventually enhance speed, but all of these changes will only be possible by testing the code in production on a sample card.

Happy to merge from my side!

iceman1001 commented:
Excellent work!
